File: Language/src/org/cgsuite/lang/InputException.java
package org.cgsuite.lang;

import java.util.ArrayList;
import java.util.List;
import org.antlr.runtime.Token;

public class InputException extends CgsuiteException
{
    private static final long serialVersionUID = 1L;

    private String invocationTarget;
    private List<Token> tokenStack;

    public InputException(String message)
    {
        super(message);
        tokenStack = new ArrayList<Token>();
    }

    public InputException(String message, Throwable e)
    {
        super(message, e);
        tokenStack = new ArrayList<Token>();
    }

    public InputException(Token token, String message)
    {
        this(message);
        addToken(token);
    }

    public List<Token> getTokenStack()
    {
        return tokenStack;
    }

    public void addToken(Token token)
    {
        if (token == null)
            throw new IllegalArgumentException("token == null");
        
        tokenStack.add(token);
    }
    
    public String getInvocationTarget()
    {
        return invocationTarget;
    }
    
    public void setInvocationTarget(String invocationTarget)
    {
        this.invocationTarget = invocationTarget;
    }
}

File: lib/core/src/main/scala/org/cgsuite/exception/InputException.scala
package org.cgsuite.exception

import org.antlr.runtime.Token
import org.antlr.runtime.tree.Tree
import org.cgsuite.lang.Node._

import scala.collection.mutable

object InputException {

  def apply(msg: String, e: Throwable = null, token: Option[Token] = None): InputException = {
    val exc = InputException(msg, e)
    exc.tokenStack ++= token
    exc
  }

  def apply(msg: String, tree: Tree): InputException = {
    InputException(msg, token = Some(tree.token))
  }

}

case class InputException private (msg: String, e: Throwable) extends CgsuiteException(msg, e) {

  var invocationTarget: Option[String] = None
  val tokenStack = mutable.MutableList[Token]()

  def addToken(token: Token): Unit = {
    tokenStack += token
  }

  def msgWithLocation: String = {
    tokenStack.headOption match {
      case None => s"[unknown location] $msg"
      case Some(t) => s"[${t.getInputStream.getSourceName} line ${t.getLine}:${t.getCharPositionInLine}] $msg"
    }
  }

}
